### DISCUSSION TITLE
Support live reload

### DISCUSSION THREAD

**Discussion by maralorn** [srid/neuron#200] [Ideas]
When you iterate over your zettels in an editor and use "rib -wS" it might be nice to have the browser live reload changes.
I use https://livejs.com/ in a similar project and that works quite nicely and simple. Of course this should be strictly optionial. The only question is where exactly to add the flag as for this to work we need to change the compilation output of the zettels.

**Comment by srid** [comment]
Related: https://github.com/srid/neuron/issues/75

**Comment by srid** [comment]
UX ideas [here](https://github.com/iamcco/markdown-preview.nvim#--markdown-preview-for-neovim--).

**Comment by maralorn** [comment]
Wow, that looks amazing. I wonder if we could throw this together with the lsp support (#213) we were talking about?

The point here is `rib -wS` watches files right? But a language server get's the code of the current file somehow transmitted (before it's saved). So if we could integrate that with "-wS" we could make a language server that continuously offers the file over http. Now, if we want to reload on nearly every keypress live.js might actually be to slow and we maybe need a websocket to trigger reloads?

Also I don‘t know if the language server protocol has a concept of focused file and scroll position.

**Comment by maralorn** [comment]
Hm, if we don't have the scroll position from the editor we can always focus the last change made.

**Comment by sureyeaah** [comment]
@maralorn Better yet we could focus on the current cursor position (which LSP supports).

**Comment by maralorn** [comment]
Uh, I didn‘t know. How great.

**Comment by srid** [comment]
A basic version of this can be implemented using [the `reload` tool](https://github.com/alallier/reload#using-reload-as-a-command-line-application).

```
nix-env -if https://github.com/srid/reload/archive/nixify.tar.gz
neuron rib -w  # Leave running in another terminal
reload -b -d ~/zettelkasten/.neuron/output
```

We basically have to replace rib's server with this one.

---

Later, once LSP support is in, we can do cool stuff like auto-scrolling to current cursor position (discussed above).

**Comment by fiatjaf** [comment]
In my opinion this is unnecessary and a distraction.
Live-reload for text rendered on a default template? It makes no sense.

**Comment by srid** [comment]
@fiatjaf As a prototype, I'd like to see something like this but possibly as a separate tool (outside of neuron). What I'd like is a browser "mirror" for my neovim session. If I have 3 split windows, for example, the browser mirror (a chrome tab) should also display 3 split windows in exact split configuration. If I scroll the neovim buffer, I'd like the browser window to scroll too (commonmark-hs's sourcepos can be used for it). The idea is to always have a rendered snapshot of your neovim view to the side.

**Comment by srid** [comment]
Demo here (it doesn't do split mirroring): https://github.com/iamcco/markdown-preview.nvim#introduction

**Comment by srid** [comment]
cf. https://github.com/srid/neuron/issues/496#issuecomment-749242607

**Comment by srid** [comment]
Using [parcel](https://parceljs.org/),

```
neuron gen -w # in other terminal
parcel .neuron/output/index.html --open
```

gives you a reloading static html server.

**Comment by srid** [comment]
Swapping out neuron's server with [ema](https://notes.srid.ca/ema-announce) will give us live reload for free.
